README.md
# MagicMirrorModule-YouLess2

## It´s a Fork from https://github.com/QNimbus/MMM-YouLess

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/). This module reads from a YouLess energy meter in your network to display realtime energy usage information.

## Installation

1. Navigate into your MagicMirror's `modules` folder and run:
```
$ git clone https://github.com/eckonator/MMM-YouLess2.git
```
1. Install the dependencies: 
```
$ cd MMM-YouLess2 && npm install --only=production
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-YouLess2',
            config: {
                // required settings
                youlessHost: "192.168.178.77",          // or youless.home.network.nl
                updateInterval: 30 * 1000,          	// every 30 seconds
                
                // optional settings
                totalGauge: 5000,                       // max watt for drawing circle
                gaugeUniqueId: "YouLess2_Gauge_1",      // must be unique, if you have multiple instances, rename the second one, e.g.: YouLess2_Gauge_2
                gaugeValueLabel: "Watt",                // gauge value label, directly after the value
                gaugeLabel: "Stromlast",                // gauge label in bottom of the gauge
                gaugeStyle: "Arch",                     // possible Options: Full | Arch | Semi
                
                // change it, if you know what you do
                initialLoadDelay: 0, 	            	// 0 seconds delay
                retryDelay: 2 * 1000,	            	// retry after 2 seconds
                gaugeStripe: 0,                         // possible Options: 0 | 1
                gaugeSize: 250,                         // size of the gauge
                gaugeColor: "#ffffff",                  // color of the active gauge
                gaugeBackColor: "rgba(255,255,255,.3)", // color of the inactive gauge
                gaugeCircleWidth: 16,                   // with of the gauge circle
                gaugeAnimationstep: 0,                  // filling gauge with or without animation
                gaugeAnimateCircleColors: false,        // animate circle colors
                gaugeAnimateTextColors: false,          // animate texte colors
                gaugeLabelColor: "#ffffff",             // label color
                gaugeTextSize: .11                      // text size
            }
        }
    ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `youlessHost`| *Required* Your YouLess ip address or hostname<br>**Default value:** 192.168.178.77<br>**Type:** `string`
| `updateInterval` | *Optional* Interval at which the module fetches new data<br><br>**Default value:** 30 * 1000 milliseconds (30 seconds)<br>**Type:** `int`(milliseconds)<br>
| `initialLoadDelay` | *Optional* Initial delay when module is first loaded<br><br>**Default value:** 0 milliseconds (0 seconds)<br>**Type:** `int`(milliseconds)<br>
| `retryDelay`     | *Optional* If no data is received, retry again after delay<br><br>**Default value:** 2 * 1000 milliseconds (2 seconds)<br>**Type:** `int`(milliseconds)<br>
| `totalGauge` | *Required* Max watt for drawing circle<br><br>**Default value:** 5000 Watt<br>**Type:** `int`(Watt)<br>
| `gaugeUniqueId` | *Required* Must be unique, if you have multiple instances, rename the second one, e.g.: YouLess2_Gauge_2<br><br>**Default value:** YouLess2_Gauge_1<br>**Type:** `string`<br>
| `gaugeValueLabel` | *Optional* Gauge value label, directly after the value<br><br>**Default value:** Watt<br>**Type:** `string`<br>
| `gaugeLabel` | *Optional* Gauge label in bottom of the gauge<br><br>**Default value:** Stromlast<br>**Type:** `string`<br>
| `gaugeStyle` | *Optional* Possible Options: Full / Arch / Semi<br><br>**Default value:** Arch<br>**Type:** `string`<br>
| `gaugeStripe` | *Optional* possible Options: 0/1<br><br>**Default value:** 0<br>**Type:** `boolean`<br>
| `gaugeSize` | *Optional* size of the gauge<br><br>**Default value:** 250<br>**Type:** `int` (pixel)<br>
| `gaugeColor` | *Optional* color of the active gauge<br><br>**Default value:** #ffffff<br>**Type:** `string`<br>
| `gaugeBackColor` | *Optional* color of the inactive gauge<br><br>**Default value:** rgba(255,255,255,.3)<br>**Type:** `string`<br>
| `gaugeCircleWidth` | *Optional* with of the gauge circle<br><br>**Default value:** 16<br>**Type:** `int` (pixel)<br>
| `gaugeAnimationstep` | *Optional* filling gauge with or without animation<br><br>**Default value:** 0<br>**Type:** `boolean`<br>
| `gaugeAnimateCircleColors` | *Optional* animate circle colors<br><br>**Default value:** false<br>**Type:** `boolean`<br>
| `gaugeAnimateTextColors` | *Optional* animate texte colors<br><br>**Default value:** false<br>**Type:** `boolean`<br>
| `gaugeLabelColor` | *Optional* label color<br><br>**Default value:** #ffffff<br>**Type:** `string`<br>
| `gaugeTextSize` | *Optional* text size<br><br>**Default value:** .11<br>**Type:** `int` (em)<br>

## Building from TypeScript source

Todo...

## Screenshots

![alt text][ss_01]

[ss_01]: public/images/module_ss_01.png "Example of YouLess2 module at work"

CHANGELOG.md
#Changlelog

### [0.1.0] - Initial commit

node_helper.js
/* Magic Mirror
 * Node Helper: MMM-YouLess2
 *
 * By M. Eckert
 * MIT Licensed.
 */

var request = require("request");
var NodeHelper = require("node_helper");

var nodeHelper = NodeHelper.create({
	config: {},

	// Override start method.
	start: function () {
	},

	// Override socketNotificationReceived method.
	socketNotificationReceived: function (notification, payload) {
		var self = this;

		var config = function (payload) {
			self.config = payload;
		};

		var query = function (payload) {
			var options = {
				uri: `http://${self.config.youlessHost}/a?f=j`,
				method: "GET",
			};

			request(options, function (error, response, body) {
				if (response && response.statusCode === 200) {
					self.sendSocketNotification("RESPONSE", body);
				}
			});
		};

		var acceptedNotifications = {
			"CONFIG": config,
			"QUERY": query,
		};

		// Call appropriate handler for socket notification
		acceptedNotifications[notification] && typeof acceptedNotifications[notification] === "function" && acceptedNotifications[notification](payload);
	},
});

module.exports = nodeHelper;

MMM-YouLess2.js
/* Magic Mirror
 * Module: MMM-YouLess2
 *
 * By M. Eckert
 * MIT Licensed.
 */

Module.register("MMM-YouLess2", {
	// Module config defaults.
	defaults: {
		// required settings
		youlessHost: "192.168.178.77",          // or youless.home.network.nl
		updateInterval: 30 * 1000,          	// every 30 seconds

		// optional settings
		totalGauge: 5000,                       // max watt for drawing circle
		gaugeUniqueId: "YouLess2_Gauge_1",      // must be unique, if you have multiple instances, rename the second one, e.g.: YouLess2_Gauge_2
		gaugeValueLabel: "Watt",                // gauge value label, directly after the value
		gaugeLabel: "Stromlast",                // gauge label in bottom of the gauge
		gaugeStyle: "Arch",                     // possible Options: Full | Arch | Semi

		// change it, if you know what you do
		initialLoadDelay: 0, 	            	// 0 seconds delay
		retryDelay: 2 * 1000,	            	// retry after 2 seconds
		gaugeStripe: 0,                         // possible Options: 0 | 1
		gaugeSize: 250,                         // size of the gauge
		gaugeColor: "#ffffff",                  // color of the active gauge
		gaugeBackColor: "rgba(255,255,255,.3)", // color of the inactive gauge
		gaugeCircleWidth: 16,                   // with of the gauge circle
		gaugeAnimationstep: 0,                  // filling gauge with or without animation
		gaugeAnimateCircleColors: false,        // animate circle colors
		gaugeAnimateTextColors: false,          // animate texte colors
		gaugeLabelColor: "#ffffff",             // label color
		gaugeTextSize: .11                      // text size
	},

	start: function () {
		this.loaded = false;
		this.powerUsage = -1;
		Log.info("Starting module: " + this.name);

		this.scheduleUpdate(this.config.initialLoadDelay);
	},

	getStyles: function () {
		return [
			this.file("public/css/styles.css"),
			"font-awesome.css",
		];
	},

	getScripts: function () {
		return [
			"https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js",  // this file will be loaded from the jquery servers.
			this.file("public/js/GaugeMeter.js"),
		];
	},

	getDom: function () {
		var wrapper = jQuery(document.createElement("div")).addClass("GaugeMeter").attr("id", this.config.gaugeUniqueId);
		var obj = {
			percent: ((this.powerUsage / this.config.totalGauge) * 100),
			size: this.config.gaugeSize,
			color: this.config.gaugeColor,
			back: this.config.gaugeBackColor,
			width: this.config.gaugeCircleWidth,
			style: this.config.gaugeStyle,
			stripe: this.config.gaugeStripe,
			animationstep: this.config.gaugeAnimationstep,
			animate_gauge_colors: this.config.gaugeAnimateCircleColors,
			animate_text_colors: this.config.gaugeAnimateTextColors,
			label: this.config.gaugeLabel,
			label_color: this.config.gaugeLabelColor,
			text: this.powerUsage,
			text_size: this.config.gaugeTextSize
		}
		wrapper.data(obj);
		wrapper.gaugeMeter();
		return wrapper[0];
	},

	updatePowerUsage: function () {
		var self = this;

		this.sendSocketNotification("QUERY");

		// Use default timeout if already loaded, else use the retryDelay if module hasn't loaded yet
		self.scheduleUpdate((self.loaded) ? -1 : self.config.retryDelay);
	},

	processPowerUsage: function (data) {
		if (data) {
			var spanSeletor = "#" + this.config.gaugeUniqueId + " span";
			var oldWatt = jQuery(spanSeletor).text();

			jQuery(spanSeletor).attr("data-after", this.config.gaugeValueLabel);

			this.powerUsageHigh = this.powerUsageHigh < data.pwr ? data.pwr : this.powerUsageHigh;
			this.powerUsage = data.pwr;
			this.loaded = true;

			jQuery("#" + this.config.gaugeUniqueId).gaugeMeter({
				percent: ((this.powerUsage / this.config.totalGauge) * 100)
			});

			this.animateValue(spanSeletor, oldWatt, this.powerUsage, 500);
		}
	},

	/* scheduleUpdate()
 	 * Schedule next update.
 	 *
 	 * argument delay number - Milliseconds before next update. If empty, this.config.updateInterval is used.
 	 */
	scheduleUpdate: function (delay) {
		var nextLoad = this.config.updateInterval;
		if (typeof delay !== "undefined" && delay >= 0) {
			nextLoad = delay;
		}

		var self = this;
		setTimeout(function () {
			self.updatePowerUsage();
		}, nextLoad);
	},

	notificationReceived: function (notification, payload, sender) {
		if (notification === "DOM_OBJECTS_CREATED") {
			this.loaded = true;
			this.sendSocketNotification("CONFIG", this.config);
		}
	},

	socketNotificationReceived: function (notification, payload) {
		var self = this;

		var responseReceived = function (payload) {
			self.processPowerUsage(JSON.parse(payload));
		};

		var acceptedNotifications = {
			"RESPONSE": responseReceived,
		};

		// Call appropriate handler for socket notification
		acceptedNotifications[notification] && typeof acceptedNotifications[notification] === "function" && acceptedNotifications[notification](payload);
	},

	animateValue: function (element, start, end, duration) {
		// assumes integer values for start and end

		var obj = jQuery(element);
		var range = end - start;
		// no timer shorter than 50ms (not really visible any way)
		var minTimer = 50;
		// calc step time to show all interediate values
		var stepTime = Math.abs(Math.floor(duration / range));

		// never go below minTimer
		stepTime = Math.max(stepTime, minTimer);

		// get current time and calculate desired end time
		var startTime = new Date().getTime();
		var endTime = startTime + duration;
		var timer;

		function run() {
			var now = new Date().getTime();
			var remaining = Math.max((endTime - now) / duration, 0);
			var value = Math.round(end - (remaining * range));
			obj.html(value);
			if (value == end) {
				clearInterval(timer);
			}
		}

		timer = setInterval(run, stepTime);
		run();
	}
});